### DISCUSSION TITLE
dart_language_server broken.

### DISCUSSION THREAD

**Discussion by zw963** [joaotavora/eglot#862] [Q&A] [answered]
```
Following is error message when working with dart-mode.

 \")\207" [#<buffer my_transactions_page.dart> (#0) eglot--managed-mode remove-hook post-command-hook nil buffer-live-p apply eglot--connect eglot--guess-contact] 4]): (file-missing "Searching for program" "No such file or directory" "dart_language_server")
```

From the dart-lang official document, dart start to support  `Dart analysis server` from version 2.2.0 , now it use form like this:

```sh
dart language-server --client-id my-editor.my-plugin --client-version 1.2
```
Please check.

https://github.com/dart-lang/sdk/blob/master/pkg/analysis_server/tool/lsp_spec/README.md
     
* Server used:      dart         
* Emacs version:      28.1       
* Operating system:     Arch Linux     
* Eglot version:    master         
* Eglot installation method:  manually      
* Using Doom:      No          

#### LSP transcript - M-x eglot-events-buffer (mandatory unless Emacs inoperable)

Debugger entered--Lisp error: (error "[eglot] Can't find an Eglot events buffer!")
  error("[eglot] %s" "Can't find an Eglot events buffer!")
  eglot--error("Can't find an Eglot events buffer!")
  eglot-events-buffer(nil)
  funcall-interactively(eglot-events-buffer nil)
  command-execute(eglot-events-buffer record)
  execute-extended-command(nil "eglot-events-buffer" "eglot-ev")
  funcall-interactively(execute-extended-command nil "eglot-events-buffer" "eglot-ev")
   
#### Minimal configuration (mandatory)

```cl
(require 'eglot_init)
(add-hook 'dart-mode-hook 'eglot-ensure)
```

**Comment by zw963** [comment]
I don't know when to use `--client-id my-editor.my-plugin` or `--client-version 1.2`, but it seem like just work after setting like this:

```cl
(setq eglot-server-programs
      `((dart-mode . ("dart" "language-server"))))
```

**Comment by zw963** [comment]
This default settings never work with recent version dart, i consider this as a issue?

**Reply by joaotavora** [reply]
Yes, but to have an issue you need to fill requirements of an issue report. You didn't :-) and that's fine.  So let's discuss this here.

In the issue report template, there is a request for a minimal configuration recipe.  Perhaps it should be called minimum reproduction recipe, meaning the minimum amount of information that any programmer that demonstrates the bug, including a clear description of what is your _expectation_ after a certain number of _reproduction steps_ followed by a clear description of what your _observation_ (presumably different to the expectation) was.  

This is not clear from the information you provided, at least not to me.  When it is (to me or to someone else) we can convert this to an issue that will have the best chance of receiving a developer's attention.

But in the meantime you can keep adding information here.

**Reply by zw963** [reply]
Sorry, original issue typo, i give my minimum config here. (with extra only load-path correct set)

```cl
(require 'dart-mode)
(add-to-list 'auto-mode-alist '("\\.dart\\'" . dart-mode))
(require 'eglot)
(add-hook 'dart-mode-hook 'eglot-ensure)
```

Following is error backtrace

```cl
Error in post-command-hook ((closure ((--cl-maybe-connect-- closure #1 nil (progn (remove-hook 'post-command-hook --cl-maybe-connect-- nil) (let ((G1 buffer)) (if (buffer-live-p G1) (save-current-buffer (set-buffer G1) (if eglot--managed-mode nil (apply #'eglot--connect (eglot--guess-contact)))))))) (buffer . #<buffer my_transactions_page.dart>) eglot--managed-mode eglot-lsp-context company-tooltip-align-annotations company-backends markdown-fontify-code-blocks-natively t) nil (progn (remove-hook 'post-command-hook --cl-maybe-connect-- nil) (let ((G1 buffer)) (if (buffer-live-p G1) (save-current-buffer (set-buffer G1) (if eglot--managed-mode nil (apply #'eglot--connect (eglot--guess-contact))))))))): (file-missing "Searching for program" "No such file or directory" "dart_language_server")
```

Following is result when run `eglot-events-buffer`

```cl
Debugger entered--Lisp error: (error "[eglot] Can't find an Eglot events buffer!")
  error("[eglot] %s" "Can't find an Eglot events buffer!")
  eglot--error("Can't find an Eglot events buffer!")
  (if buffer (display-buffer buffer) (eglot--error "Can't find an Eglot events buffer!"))
  (let ((buffer (if server (jsonrpc-events-buffer server) (cl-find "\\*EGLOT.*events\\*" (buffer-list) :key #'buffer-name :test #'string-match)))) (if buffer (display-buffer buffer) (eglot--error "Can't find an Eglot events buffer!")))
  eglot-events-buffer(nil)
  funcall-interactively(eglot-events-buffer nil)
  command-execute(eglot-events-buffer record)
  execute-extended-command(nil "eglot-events-buffer" nil)
```

## the work around (test works on dart 2.16.1)

```cl
(require 'dart-mode)
(add-to-list 'auto-mode-alist '("\\.dart\\'" . dart-mode))
(require 'eglot)
(setq eglot-server-programs (remove '(dart-mode "dart_language_server") eglot-server-programs))
(add-to-list 'eglot-server-programs '(dart-mode . ("dart" "language-server")))
(add-hook 'dart-mode-hook 'eglot-ensure)
```

**Reply by zw963** [reply]
Any update on this?

**Reply by leungbk** [reply]
It looks like the existing `dart` command in `eglot.el` corresponds to an old, unmaintained language server, different from the official one you're linking to. So it does seem like you'll have to PR with a command for the official LS.

I don't use Dart, so I don't know what significance, if any, the `client-id` stuff holds. Can you figure this out from grepping through their server code?

**Reply by ericdallo** [reply]
client-id is just a analytics thing, we pass it as `emacs.lsp-dart` at [lsp-dart](https://github.com/emacs-lsp/lsp-dart/blob/master/lsp-dart.el#L119)

**Reply by nemethf** [reply]
Ideally, the server should deduce this kind of thing from the clientInfo.  I'm not sure if dart reads the clientInfo.  Eglot does not currently send it, anyway: #818.

**Reply by bradyt** [reply]
I can reproduce the issue above. Let me know if you're having trouble reproducing it. In a sense, there is a fix, with `dart pub global activate dart_language_server`.

But Eglot's README points to https://github.com/natebosch/dart_lsp, which indicates that `dart_language_server` is deprecated, and links to https://github.com/dart-lang/sdk/blob/master/pkg/analysis_server/tool/lsp_spec/README.md.

What is currently preventing this discussion from being converted back to an issue?

There seems to be hesitation about the additional flags. Considering the phrase:

> Start the language server using the `dart language-server` command. Pass the `--client-id` and `--client-version` flags to identify your editor/plugin and version:

This could be interpreted that we should use `dart language-server`. Then, optionally, identify the editor/plugin and version. A user could always add the flags themselves in their init file. I'm not sure considering the flags should block on fixing the issue by updating the server.

I confirm, the following is working for me:

```elisp
(with-eval-after-load 'eglot
  (add-to-list 'eglot-server-programs
               '(dart-mode . ("dart" "language-server"))))
```

**Reply by joaotavora** [reply]
> I confirm, the following is working for me:

If that is known to work and you want to make that the default in Eglot, submit a pull request.

> What is currently preventing this discussion from being converted back to an issue?

How would that help? Issues in Eglot are meant to start with formal descriptions, almost like programs, of effects that I (or other Eglot devs) can mechanically reproduce with no domain knowledge, and little or no investigation into the domain ("dart" things, in this case).  Everything else should be a discussion, where domain specific hints (like yours) are more than welcome.  This seems to be a dart LS issue.

**Reply by zw963** [reply]
Sure, i can, but, if i need do somethings like this first?

```markdown
Eglot is subject to the same [copyright assignment](https://www.fsf.org/licensing/contributor-faq) policy as GNU Emacs.

Any [legally significant](https://www.gnu.org/prep/maintain/html_node/Legally-Significant.html#Legally-Significant) contributions can only be merged after the author has completed their paperwork. Please ask for the request form, and we'll send it to you.
```

**Reply by zw963** [reply]
PR https://github.com/joaotavora/eglot/pull/914

**Reply by joaotavora** [reply]
Pretty sure the change in question is _legally_ insignificant :-D

**Reply by joaotavora** [reply]
pushed it to master.

**Comment by joaotavora** [comment, marked as answer]
The default `dart-mode` entry in `eglot-server-programs` should now fix the issues described.
